lib/ecto_diff.ex

```diff
@@ -303,37 +303,47 @@ defmodule EctoDiff do
   defp diff_association(previous, current, %{cardinality: :many, field: field, related: struct}, acc) do
     primary_key_fields = struct.__schema__(:primary_key)
 
-    if primary_key_fields == [],
-      do: raise("cannot determine difference in many association with no primary key for `#{struct}`")
-
-    {previous_map, keys} =
-      Enum.reduce(previous, {%{}, []}, fn x, {map, keys} ->
-        key = Map.take(x, primary_key_fields)
-        {Map.put(map, key, x), [key | keys]}
-      end)
-
-    {current_map, keys} =
-      Enum.reduce(current, {%{}, keys}, fn x, {map, keys} ->
-        key = Map.take(x, primary_key_fields)
-        {Map.put(map, key, x), [key | keys]}
-      end)
-
-    keys = keys |> Enum.reverse() |> Enum.uniq()
-
     diffs =
-      keys
-      |> Enum.map(fn key ->
-        prev_child = Map.get(previous_map, key)
-        current_child = Map.get(current_map, key)
+      if primary_key_fields == [] do
+        from = max(length(previous), length(current)) - 1
 
-        do_diff(prev_child, current_child)
-      end)
-      |> Enum.reject(&no_changes?/1)
+        from
+        |> Range.new(0)
+        |> Enum.map(fn i ->
+          prev_child = Enum.at(previous, i)
+          current_child = Enum.at(current, i)
 
-    if diffs == [] do
-      acc
-    else
-      [{field, diffs} | acc]
+          do_diff(prev_child, current_child)
+        end)
```

Comment:
Along with doughsay's suggestion, once you bring in that `zip_longest` function, this implementation becomes:

```ex
zip_longest(previous, current)
|> Enum.map(fn
  {[], current_child} -> do_diff(nil, current_child)
  {prev_child, []} -> do_diff(prev_child, nil)
  {prev_child, current_child} -> do_diff(prev_child, current_child)
end)
```

Alternatively, you could alter the `zip_longest` functions to fill `nil` instead of an empty list; either would work and would simplify your impl here down to

```ex
zip_longest(previous, current)
|> Enum.map(&do_diff(&1, &2))
```